content/dharma/abstract-data-structures.md

```diff
@@ -0,0 +1,118 @@
+---
+Title: "Introduction to ADTs in JavaScript"
+Description: "An introduction to Abstract Data Types"
+DatePublished: "2022-04-25"
+Categories:
+  - "code-foundations"
+  - "computer-science"
+Tags:
+  - "Data Structures"
+  - "Classes"
+  - "OOP"
+CatalogContent:
+  - "paths/computer-science"
+  - "introduction-to-javascript"
+---
+
+![Introduction to Abstract Data Type](./adt.png)
+
+## Introduction
+
+An Abstract Data Type, as the name suggests, is an abstract understanding of a data structure. An **Abstract Data Type** is defined through its behavior and characteristics, particularly in terms of what data can be stored into it, the operations that can be performed on this data, and the behavior of these operations. For example, stacks and queues can be internally implemented using linked-lists made up of nodes or lists/arrays. However, the primary function of a stack is to be a Last In First Out(LIFO) data structure and the primary function of a queue is to be a First In First Out(FIFO) data structure. The behavior, from the point of the user, remains intact, regardless of the internal implementation either using linked-lists or arrays/lists. If the user was interacting with a stack, the user will simply worry about pushing data onto the stack or popping data off the stack. The user won't need to have the knowledge how that stack is working internally.
+
+In contrast to the data structures which are specific and detailed implementations that deal with how the data structure does its job, abstract data type focuses on _what it does_ and not how it does its job. In short, the ADT defines what that particular data construct must do and the data structure is the concrete implementation of that construct.
+
+An analogy to explain abstract data type in terms of web development would be CRUD APIs. The user of any CRUD API has to simply know what request method(GET, POST, etc.) should they send and if they followed the rules of the API, the API server would send data back. The user didn't have to worry about the internal workings of the API server. The user simply had to know the rules of interactions and behavior of CRUD API. In this case, the CRUD API is functioning as an abstract data type _from the perspective of the user_.
+
+There are no specific rules which force the implementation of particular methods and operations in a particular ADT. This is decided based on the requirements in a use-case scenario and therefore, ultimately a purely design decision.
+
+## Why use ADTs?
```

Comment:
Up to now the usage of ADT or abstract data type has been inconsistent. You should probably stick to the ADT acronym once it's introduced.

Reply:
I'll introduce the "Abstract Data Type" as ADT in the first line and all the succeeding mentions will only mention ADT from that point.